pkg/querier/batch/stream.go

```diff
@@ -39,64 +45,125 @@ func (bs *batchStream) atTime() int64 {
 }
 
+func (bs *batchStream) atFloatHistogram() (int64, *histogram.FloatHistogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.FloatHistogramValues[b.Index]
+}
+
+// mergeStreams merges streams of Batches of the same series over time.
+// Samples are simply merged by time when they are the same type (float/histogram/...), with the left stream taking precedence if the timestamps are equal.
+// When sample are different type, batches are not merged. In case of equal timestamps, histograms take precedence since they have more information.
 func mergeStreams(left, right batchStream, result batchStream, size int) batchStream {
+
 	// Reset the Index and Length of existing batches.
 	for i := range result {
 		result[i].Index = 0
 		result[i].Length = 0
 	}
+
 	resultLen := 1 // Number of batches in the final result.
 	b := &result[0]
 
-	// This function adds a new batch to the result
-	// if the current batch being appended is full.
-	checkForFullBatch := func() {
+	// Step to the next Batch in the result, create it if it does not exist
+	nextBatch := func(valueTypes chunkenc.ValueType) {
+		// The Index is the place at which new sample
+		// has to be appended, hence it tells the length.
+		b.Length = b.Index
+		resultLen++
+		if resultLen > len(result) {
+			// It is possible that result can grow longer
+			// then the one provided.
+			result = append(result, createBatch(valueTypes))
+		}
+		b = &result[resultLen-1]
+	}
+
+	// ensureBatchType sets the Batch type and returns true if it was successful
+	ensureBatchType := func(valueTypes chunkenc.ValueType) bool {
+		if b.ValueTypes == valueTypes {
+			return true
+		}
+		if b.ValueTypes == chunkenc.ValNone {
+			// Uninitialized Batch
+			initBatchType(b, valueTypes)
+			return true
+		}
+		// These should be rare, means that we're merging into something where the chunk type changed
+		if b.Index == 0 {
+			// Batch not written yet
+			result[resultLen-1] = createBatch(valueTypes)
+			b = &result[resultLen-1]
+			return true
+		}
+		// Batch already started, finish and start new
+		// We don't know it's type so return false (but it will either be None or have b.Index==0)
+		nextBatch(valueTypes)
+		return false
+	}
+
+	// Ensure that the batch we're writing to is not full and of the right type.
+	ensureBatch := func(valueTypes chunkenc.ValueType) {
 		if b.Index == size {
-			// The batch reached it intended size.
+			// The batch reached its intended size.
 			// Add another batch the the result
 			// and use it for further appending.
-
-			// The Index is the place at which new sample
-			// has to be appended, hence it tells the length.
-			b.Length = b.Index
-			resultLen++
-			if resultLen > len(result) {
-				// It is possible that result can grow longer
-				// then the one provided.
-				result = append(result, chunk.Batch{})
-			}
-			b = &result[resultLen-1]
+			nextBatch(valueTypes)
+		}
+		if !ensureBatchType(valueTypes) {
+			ensureBatchType(valueTypes)
 		}
```

Comment:
I think this should do the same? Also it seems that this is the only usage of `ensureBatchType` so the return value isn't considered.
```suggestion
		ensureBatchType(valueTypes)
```

Reply:
It's a bit more complicated. There's a special case where upon noticing that the Batch type is wrong, we need to close the current Batch (call nextBatch) and then check again that the Batch is of the correct type. So basically run the algorithm twice, but on the second try we can be sure that we're not in the middle of a Batch. I'll try to make this nicer.